design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,207 @@
+
+## Update/Rollback Compatibility
+
+* SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten. 
+* UI must be capable of detecting boot sources for DataVolumeTemplates with sources for both VolumeReferences and PVCs
+
+## Implementation Phases
+
+**Phase1** - CDI APIs
+* VolumeReference API/controller
+* VolumeImportPoller API/controller
+
+**Phase2** - SSP + Common Templates
+* Common Templates transition to VolumeReferences for DataVolumeTemplate source
+* SSP Operator creates VolumeImportPollers for automatic PVC import
+
+**Phase3** - HCO and UI
+* HCO Team integrates with SSP Operator’s API for setting volume import sources
+* UI Team transitions to detecting boot source using a linked VolumeReference rather than looking directly for a PVC of a specific name.
+
+
+
+# API Examples
+## VM + DataVolumeTemplate + VolumeReference as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+   pvc:
+     accessModes:
+     - ReadWriteOnce
+     resources:
+       requests:
+         storage: 2Gi
+   source:
+     volumeReference:
+       name: fedora
+       namespace: some-namespace
+```
+
+## VolumeImportPoller Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the VolumeImportPoller will import them as new PVCs and automatically manage updating the corresponding VolumeReference.
+
+* **source** specifies where to poll from
+* **pollInterval** specifies how often to look for new imports 
+* **garbageCollectOutdated** specifies whether old PVCs should be cleaned up after a new PVC is imported.
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportPoller
+metadata:
+   name: fedora-poller
+spec:
+   source:
+     registry:
+       url: docker://quay.io/kubevirt/fedora
+    pollInterval: 12h
+    garbageCollectOutdated: true
+status:
+  managedVolumeReference:
+      name: fedora
+```
+
+## VolumeReference Example
+
+This example shows how a PVC is linked using the claimName field on the VolumeReference spec. The VolumeReference’s status section will indicate if the claimName is linked or dangling. 
+
+The UI can use the phase field to determine if a boot source exists for a VolumeReference or not. 
+
+* **Phase: linked** indicates that the the claimName is set and that the corresponding PVC exists and is populated
+* **Phase: dangling** indicates that the claimName points to a non-existent PVC, meaning the underlying PVC has either been deleted or is not created yet. 
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeReference
+metadata:
+   name: fedora
+spec:
+    source:
+        pvc:
+            namespace: somenamespace
```

Comment:
As we have namespace here, but not in the poller.status, but also in the DV.
What is the overall namespace relationship?
From what I read so far it's:
- DV and PVC do not have to be in the same NS
- PVC and VolumeRef do not have to be in the same NS
- VolumeRef and VolumePoller have to be in the same NS

Is this correct, intended?

Reply:
> PVC and VolumeRef do not have to be in the same NS

Strictly speaking, the PVC and VolumeRef don't have to be in the same namespace. However, in practice they always will be when these pipelines are being used with the volume import poller. 

> Is this correct, intended?

Everything involved with the image pipeline is in the same Namespace:
-VolumePoller
-VolumeReference
-Imported golden image PVCs

A DataVolume in any namespace can then use a VolumeReference in the golden image namespace to import a PVC from the golden image namespace.

Reply:
Just want to make sure we're consistent with where we allow to bridge namespaces.